src/drivers/flash/flash.c

```diff
@@ -1,328 +1,132 @@
-		return -1;
-	}
-	dev_name += sizeof("mtd");
-
-	if(!isdigit((int)dev_name[0])) {
-		return -1;
-	}
-
-	sscanf(dev_name, "%d", &idx);
+static struct flash_dev *flashdev_tab[MAX_DEV_QUANTITY];
 
-	return idx;
-}
-
-struct flash_dev *flash_create(char *path, size_t size) {
+struct flash_dev *flash_alloc(void) {
 	struct flash_dev *flash;
 	int idx;
 
 	if (NULL == (flash = pool_alloc(&flash_pool))) {
 		return err_ptr(ENOMEM);
 	}
 
-	if (0 > (idx = block_dev_named(path, &flash_idx))) {
+	idx = index_alloc(&flash_idx, INDEX_MIN);
+	if (idx == INDEX_NONE) {
 		pool_free(&flash_pool, flash);
-		return err_ptr(-idx);
+		return err_ptr(ENOMEM);
 	}
 
-	flash->bdev = block_dev_create(path, &flashbdev_pio_driver, flash);
-	if (NULL == flash->bdev) {
-		index_free(&flash_idx, idx);
-		pool_free(&flash_pool, flash);
-		return err_ptr(EIO);
-	}
+	memset(flash, 0, sizeof(*flash));
+	flash->idx = idx;
 
-	flash->bdev->size = size;
+	flashdev_tab[idx] = flash;
 
 	return flash;
 }
 
-struct flash_dev *flash_get_param(char *path) {
-	struct path flash_node;
-	struct nas *nas;
-	struct node_fi *node_fi;
-
-	vfs_lookup(path, &flash_node);
-
-	if (NULL == flash_node.node) {
-		return NULL;
-	}
-	nas = flash_node.node->nas;
-	node_fi = nas->fi;
-	return (struct flash_dev *) block_dev(node_fi->privdata)->privdata;
-}
-
-int flash_delete(const char *name) {
-	struct path flash_node;
-	struct flash_dev *flash;
-	struct nas *nas;
-	struct node_fi *node_fi;
-	int idx;
+int flash_free(struct flash_dev *dev) {
+	assert(dev);
 
-	vfs_lookup(name, &flash_node);
+	index_free(&flash_idx, dev->idx);
+	flashdev_tab[dev->idx] = 0;
+	pool_free(&flash_pool, dev);
 
-	if (NULL == flash_node.node) {
-		return -1;
-	}
-
-	nas = flash_node.node->nas;
-	node_fi = nas->fi;
-	if (NULL != (flash = (struct flash_dev *)
-							block_dev(node_fi->privdata)->privdata)) {
-		if (-1 != (idx = flash_get_index((char *)name))) {
-			index_free(&flash_idx, idx);
-		}
-		pool_free(&flash_pool, flash);
-		block_dev_destroy (node_fi->privdata);
-		vfs_del_leaf(flash_node.node);
-	}
 	return 0;
 }
 
-static int flashbdev_read(struct block_dev *bdev,
-		char *buffer, size_t count, blkno_t blkno) {
-	struct flash_dev *flash;
-	uint32_t startpos, endpos;
-
-	flash = (struct flash_dev *)bdev->privdata;
-	startpos = flash->start + (blkno * flash->block_info.block_size);
-	endpos = startpos + count - 1;
-
-	if ( startpos < flash->start ) {
-		return -EINVAL;
-	}
-	if ( endpos > flash->end ) {
-		return -EINVAL;
-	}
-
-	if((!flash->drv) || (!flash->drv->flash_read)) {
-		return -EINVAL;
-	}
-	return flash->drv->flash_read(flash, startpos, buffer, count);
-}
-
-
-static int flashbdev_write(struct block_dev *bdev,
-		char *buffer, size_t count, blkno_t blkno) {
-	struct flash_dev *flash;
-	uint32_t startpos, endpos;
-
-	flash = (struct flash_dev *)bdev->privdata;
-	startpos = flash->start + (blkno * flash->block_info.block_size);
-	endpos = startpos + count - 1;
+static int flash_initialized = 0;
```

Comment:
May be use UNIT_INIT?

Reply:
This is necessary due to broken RunLevel sequence: sometimes we have to init devices before init of this module.